src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+                }
+            }
+        }
+
+        internal static void AddScaleU(float scale, float[] src, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                fixed (float* pdst = dst)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        srcVector = Sse.Multiply(srcVector, scaleVector);
+                        dstVector = Sse.Add(dstVector, srcVector);
+                        Sse.Store(pDstCurrent, dstVector);
+
+                        pDstCurrent += 4;
+                        pSrcCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                        srcVector = Sse.MultiplyScalar(srcVector, scaleVector);
+                        dstVector = Sse.AddScalar(dstVector, srcVector);
+                        Sse.StoreScalar(pDstCurrent, dstVector);
+
+                        pDstCurrent++;
+                        pSrcCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static void AddScaleSU(float scale, float[] src, int[] idx, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                fixed (int* pidx = idx)
+                fixed (float* pdst = dst)
+                {
+                    float* pSrcCurrent = psrc;
+                    int* pIdxCurrent = pidx;
+                    float* pDstCurrent = pdst;
+                    int* pEnd = pidx + count;
+
+                    while (pIdxCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Load4(pDstCurrent, pIdxCurrent);
+
+                        srcVector = Sse.Multiply(srcVector, scaleVector);
+                        dstVector = Sse.Add(dstVector, srcVector);
+                        Store4(dstVector, pDstCurrent, pIdxCurrent);
+
+                        pIdxCurrent += 4;
+                        pSrcCurrent += 4;
+                    }
+
+                    while (pIdxCurrent < pEnd)
+                    {
+                        pDstCurrent[*pIdxCurrent] += scale * *pSrcCurrent;
```

Comment:
nit: please add parenthesis where it improves readability.

You are mixing multiplication and pointers here, so `scale * *pSrcCurrent` makes you have to stop and manually process the operands apart.

Reply:
Very much agree.  Thank you!